flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/UnboundedEventTimeOverProcessFunction.scala

```diff
@@ -0,0 +1,181 @@
+  * @param forwardedFieldCount the input fields count
+  * @param intermediateType the intermediate row tye which the state saved
+  * @param inputType the input row tye which the state saved
+  *
+  */
+class UnboundedEventTimeOverProcessFunction(
+    private val aggregates: Array[AggregateFunction[_]],
+    private val aggFields: Array[Int],
+    private val forwardedFieldCount: Int,
+    private val intermediateType: TypeInformation[Row],
+    private val inputType: TypeInformation[Row])
+  extends ProcessFunction[Row, Row]{
+
+  Preconditions.checkNotNull(aggregates)
+  Preconditions.checkNotNull(aggFields)
+  Preconditions.checkArgument(aggregates.length == aggFields.length)
+
+  private var output: Row = _
+  private var accumulatorState: ValueState[Row] = _
+  private var rowState: ListState[Tuple2[Long, Row]] = _
+
+
+  override def open(config: Configuration) {
+    output = new Row(forwardedFieldCount + aggregates.length)
+    val valueSerializer: TypeSerializer[Row] =
+      intermediateType.createSerializer(getRuntimeContext.getExecutionConfig)
+    val stateDescriptor: ValueStateDescriptor[Row] =
+      new ValueStateDescriptor[Row]("accumulatorstate", valueSerializer)
+    accumulatorState = getRuntimeContext.getState[Row](stateDescriptor)
+
+    val tupleSerializer: TypeSerializer[Tuple2[Long, Row]] =
+      (new TupleTypeInfo(BasicTypeInfo.LONG_TYPE_INFO, inputType)).createSerializer(
+        getRuntimeContext.getExecutionConfig).asInstanceOf[TypeSerializer[Tuple2[Long, Row]]]
+    val tupleStateDescriptor: ListStateDescriptor[Tuple2[Long, Row]] =
+      new ListStateDescriptor[Tuple2[Long, Row]]("rowliststate", tupleSerializer)
+    rowState = getRuntimeContext.getListState[Tuple2[Long, Row]](tupleStateDescriptor)
+
+  }
+
+  /**
+    * Process one element from the input stream, not emit the output
+    *
+    * @param value The input value.
+    * @param ctx   The ctx to register timer or get current time
+    * @param out   The collector for returning result values.
+    *
+    */
+  override def processElement(
+     input: Row,
+     ctx:  ProcessFunction[Row, Row]#Context,
+     out: Collector[Row]): Unit = {
+
+    // discard later record
+    if (ctx.timestamp() >= ctx.timerService().currentWatermark()) {
+      // ensure every key just register on timer
+      ctx.timerService.registerEventTimeTimer(ctx.timerService.currentWatermark + 1)
+
+      rowState.add(new Tuple2(ctx.timestamp, input))
+    }
+  }
+
+  /**
+    * Called when a timer set fires, sort current records according the timestamp
+    * and emit the output
+    *
+    * @param timestamp The timestamp of the firing timer.
+    * @param ctx       The ctx to register timer or get current time
+    * @param out       The collector for returning result values.
+    */
+  override def onTimer(
+      timestamp: Long,
+      ctx: ProcessFunction[Row, Row]#OnTimerContext,
+      out: Collector[Row]): Unit = {
+
+    var rowList = rowState.get.iterator
```

Comment:
check immediately if `rowList.hasNext()` and leave the function if `rowList` is empty. This function will be called very often with empty lists so we should keep the overhead as low as possible.